### ISSUE TITLE
Cotton Sound

### ISSUE THREAD

**Issue by arcadez2003** [libretro/mame2003-plus-libretro#1375]
Well i cant get the newer rewritten upd7759 to play nice in the ole system16.c driver the music and sfx is fine but anything
via the upd is silent, the update is pretty simple for all of the drivers connected to that soundcore but as things tend to
go the changes in segas16b are more complex.

When i did this update on the xbox it wasn't a problem as i had the segas16b driver to begin with and i simply created
a seperate standalone version of the newer upd7759 which i use for sgeas16b and i kept the original soundcore for all
the other drivers.

Why did i do that.?? well beacuse im lazy and couldn't be bothered to test 100's of games afterwards :) but i got the
result i was looking for namely Cotton playable with full sound basically.

Anyway as it stands the sound problems will remain for Cotton we have several options i suppose.......

A) update the upd7759 and hopefully suss how to hook it up correctly to all the drivers affected and in time test all
the games connected to this to make sure there is no breakage.

B) create a sound banking hack which will hopefully fix the sound in Cotton

Alien Syndrome has one for use with the older version of the upd7759 without it the sound is busted, currently we
have Cotton using it and the sound and music work to an extent before cutting out maybe with the correct banking
address calls that will sort it.??

C) if all else fails we could always just do what i did create a standalone version of the newer upd7759 and use it
for just Cotton.

There are always multiple paths to the same destination :)

Here is link to MAME92 on the git which makes it easy to track all the changes with regards to the soundcore and the
drivers and src files affected
https://github.com/mamedev/historic-mame/commit/1cd2b6280ac28fbcdc4ee590a8c0c909a06a851c#diff-c2c7e1250cec50a00f8907735f750aa4040f2493e5721a693ce1051fedfa6193

**Comment by arcadez2003** [comment]
Here is option C the lazy option :)

As things stand music and sfx work if you coin up and start but leave it to cycle the attract mode then no sound
equally do an ingame reset and no sound, no UPD7759 sounds whatsoever as yet check the banking upd7759b_control_w
in segas16b i've like made a mistake adding the code to the ole driver although it looks correct to me

[upd7759b for Cotton.zip](https://github.com/libretro/mame2003-plus-libretro/files/8882388/upd7759b.for.Cotton.zip)

**Comment by MistyDreams** [comment]
To be honest I havent played this game before it does play nice though! ill give the ole sounds a spin. I had planned to do some more coding than I done today but the ole ball in chain had other ideas. Great work on  doing the sound wasnt expecting anything for a while to be honest. I wondering if it would be worth adding the original afighers cause i cant be arsed doning the analog controls haha

**Comment by MistyDreams** [comment]
still getting them unmapped memory writes (8000- F7FF )  no extra sounds maybe its just me that missed something i did add it to my branch

**Comment by MistyDreams** [comment]
even without this fix and with it the only music i get is on the title screen the its all stops everything once i coin upand  start the game

**Comment by arcadez2003** [comment]
Ah i forgot to remove this call from the cotton driver init  sys16_alienfix = 1; try removing it or maybe your mem maps
are different.?? anyway here is the driver im using should you wanna check it.

[sys16 72.zip](https://github.com/libretro/mame2003-plus-libretro/files/8883895/sys16.72.zip)

**Comment by arcadez2003** [comment]
Daft question you did add the upd7759b calls to your system16.c driver.?? as i dont see em in this commit....
https://github.com/MistyDreams/mame2003-plus-libretro/commit/cc5bc7558cb21a66c863affbca8d6b29a4286b48

**Comment by MistyDreams** [comment]
ill  re check thought i added it all had a moaning woman wanting to go to garden center yesterday and what did she get nothing! you right sound indeed is working sorry about that.

**Comment by MistyDreams** [comment]
looks like bullet might need this fix as well.  I done the inputs for bullet and fixed the dip switch reading order was returning the wrong order in the standard io r. Will sort cottons inputs next going to watch some tv for a bit.

**Comment by arcadez2003** [comment]
This still needs some work there are no UPD7759 sounds in Cotton and likely any other games you hookup will be the same 
it's likely something i missed in segas16b, as for Cotton inputs and dips i have those done and confirmed to be working
you can pull em from this if you want......

static MEMORY_READ16_START( cotton_readmem )
	{ 0x000000, 0x0fffff, MRA16_ROM },
	{ 0x100000, 0x100003, SYS16_MRA16_EXTRAM3 },
	{ 0x200000, 0x203fff, MRA16_RAM },
	{ 0x300000, 0x300fff, SYS16_MRA16_SPRITERAM },
	{ 0x3f0000, 0x3fffff, SYS16_MRA16_EXTRAM },
	{ 0x400000, 0x40ffff, SYS16_MRA16_TILERAM },
	{ 0x410000, 0x410fff, SYS16_MRA16_TEXTRAM },
	{ 0x500000, 0x500fff, SYS16_MRA16_PALETTERAM },
	{ 0x601002, 0x601003, input_port_0_word_r }, // player1
	{ 0x601004, 0x601005, MRA16_NOP },
	{ 0x601006, 0x601007, input_port_1_word_r }, // player2
	{ 0x601000, 0x601001, input_port_2_word_r }, // service
	{ 0x602002, 0x602003, input_port_3_word_r }, // dip1
	{ 0x602000, 0x602001, input_port_4_word_r }, // dip2
	{ 0xffc000, 0xffffff, SYS16_MRA16_WORKINGRAM },
MEMORY_END

static MEMORY_WRITE16_START( cotton_writemem )
	{ 0x000000, 0x0fffff, MWA16_ROM },
	{ 0x100000, 0x100003, SYS16_MWA16_EXTRAM3 },
	{ 0x200000, 0x203fff, MWA16_RAM },
	{ 0x300000, 0x300fff, SYS16_MWA16_SPRITERAM },
	{ 0x3f0000, 0x3fffff, SYS16_MWA16_EXTRAM },
	{ 0x400000, 0x40ffff, SYS16_MWA16_TILERAM },
	{ 0x410000, 0x410fff, SYS16_MWA16_TEXTRAM },
	{ 0x500000, 0x500fff, SYS16_MWA16_PALETTERAM },
    { 0xc40000, 0xc40001, sys16_coinctrl_w }, // Aurail hookup for now i assume this game needs this ??
	{ 0xff0006, 0xff0007, sound_command_w }, // alien sydrome hack for the UPD7759 needed or no sound 
	{ 0xfe0020, 0xfe003f, MWA16_NOP }, // Aurail config regs needed.??
	{ 0xffc000, 0xffffff, SYS16_MWA16_WORKINGRAM },
MEMORY_END

INPUT_PORTS_START( cotton )
	SYS16_JOY1
	SYS16_JOY2
	SYS16_SERVICE
	SYS16_COINAGE

PORT_START	/* DSW1 */
	PORT_DIPNAME( 0x01, 0x00, DEF_STR( Demo_Sounds ) )
	PORT_DIPSETTING(    0x01, DEF_STR( Off ) )
	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
	PORT_DIPNAME( 0x02, 0x02, DEF_STR( Unknown ) )
	PORT_DIPSETTING(    0x02, DEF_STR( Off ) )
	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
	PORT_DIPNAME( 0x04, 0x04, DEF_STR( Unknown ) )
	PORT_DIPSETTING(    0x04, DEF_STR( Off ) )
	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
	PORT_DIPNAME( 0x06, 0x06, DEF_STR( Lives ) )
	PORT_DIPSETTING(    0x04, "2" )
	PORT_DIPSETTING(    0x06, "3" )
	PORT_DIPSETTING(    0x02, "4" )
	PORT_DIPSETTING(    0x00, "5" )
	PORT_DIPNAME( 0x08, 0x08, DEF_STR( Unknown ) )
	PORT_DIPSETTING(    0x08, DEF_STR( Off ) )
	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
	PORT_DIPNAME( 0x10, 0x10, DEF_STR( Unknown ) )
	PORT_DIPSETTING(    0x10, DEF_STR( Off ) )
	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
	PORT_DIPNAME( 0x18, 0x18, DEF_STR( Difficulty ) )
	PORT_DIPSETTING(    0x10, "Easy" )
	PORT_DIPSETTING(    0x18, "Normal" )
	PORT_DIPSETTING(    0x08, "Hard" )
	PORT_DIPSETTING(    0x00, "Hardest" )
	PORT_DIPNAME( 0x20, 0x20, DEF_STR( Unknown ) )
	PORT_DIPSETTING(    0x20, DEF_STR( Off ) )
	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
	PORT_DIPNAME( 0x40, 0x40, DEF_STR( Unknown ) )
	PORT_DIPSETTING(    0x40, DEF_STR( Off ) )
	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
	PORT_DIPNAME( 0x80, 0x80, DEF_STR( Unknown ) )
	PORT_DIPSETTING(    0x80, DEF_STR( Off ) )
	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
INPUT_PORTS_END

**Comment by arcadez2003** [comment]
Notice the romboard types the banking is different for certain ones at the moment we're using a stripped down
version just for Cotton...........

static WRITE8_HANDLER( upd7759b_control_w )
{
	int size = memory_region_length(REGION_CPU2) - 0x10000;
	if (size > 0)
	{
		int bankoffs = 0;
		
		/* it is important to write in this order: if the /START line goes low
		   at the same time /RESET goes low, no sample should be started */
		upd7759b_start_w(0, data & 0x80);
		upd7759b_reset_w(0, data & 0x40);
		
		/* banking depends on the ROM board */
		switch (rom_board)
		{
			case ROM_BOARD_171_5358:
			case ROM_BOARD_171_5358_SMALL:
				/*
					D5 : /CS for ROM at A11
					D4 : /CS for ROM at A10
					D3 : /CS for ROM at A9
					D2 : /CS for ROM at A8
					D1 : A15 for all ROMs (Or ignored for 27256's)
					D0 : A14 for all ROMs
				*/
				if (!(data & 0x04)) bankoffs = 0x00000;
				if (!(data & 0x08)) bankoffs = 0x10000;
				if (!(data & 0x10)) bankoffs = 0x20000;
				if (!(data & 0x20)) bankoffs = 0x30000;
				bankoffs += (data & 0x03) * 0x4000;
				break;
			
			case ROM_BOARD_171_5521:
			case ROM_BOARD_171_5704_EXTRA_RAM:
				/*
					D5 : Unused
					D4 : Unused
					D3 : ROM select 0=A11, 1=A12
					D2 : A16 for all ROMs
					D1 : A15 for all ROMs
					D0 : A14 for all ROMs
				*/
				bankoffs = ((data & 0x08) >> 3) * 0x20000;
				bankoffs += (data & 0x07) * 0x4000;
				break;
			
			case ROM_BOARD_171_5797:
				/*
					D5 : Unused
					D4 : A17 for all ROMs
					D3 : ROM select 0=A11, 1=A12
					D2 : A16 for all ROMs
					D1 : A15 for all ROMs
					D0 : A14 for all ROMs
				*/
				bankoffs = ((data & 0x08) >> 3) * 0x40000;
				bankoffs += ((data & 0x10) >> 4) * 0x20000;
				bankoffs += (data & 0x07) * 0x04000;
				break;
		}
		cpu_setbank(1, memory_region(REGION_CPU2) + 0x10000 + (bankoffs % size));
	}
}

I thought this would be enough but maybe not.?? search for rom_board in your segas16b driver and you'll see there is
alotta code connected to the UPD7759 banking calls i had to leave out

**Comment by MistyDreams** [comment]
will need looked into but looks like the alienfix should work for bullet!

**Comment by arcadez2003** [comment]
Aye i used that for Bullet and the game seems fine soundwise

**Comment by MistyDreams** [comment]
yea the sound seems fine getting unmapped memory reads though from the cpu1

**Comment by arcadez2003** [comment]
Yip that is the sound rom loading i have i noticed this right enough......

#define ROM_BOARD_171_5521		(2)		/* 171-5521 */
#define ROM_BOARD_171_5704		(2)		/* 171-5704 - don't know any diff between this and 171-5521 */

Cotton uses this generic_5704 driver init call but there is no hookup for that in the upd7759b_control_w 

			case ROM_BOARD_171_5521:
			case ROM_BOARD_171_5704_EXTRA_RAM:
				/*
					D5 : Unused
					D4 : Unused
					D3 : ROM select 0=A11, 1=A12
					D2 : A16 for all ROMs
					D1 : A15 for all ROMs
					D0 : A14 for all ROMs
				*/
				bankoffs = ((data & 0x08) >> 3) * 0x20000;
				bankoffs += (data & 0x07) * 0x4000;
				break;

Although i assume since these calls share the same case values in this instance (2) the banking above is correct for Cotton.??

#define ROM_BOARD_171_5521		(2)		/* 171-5521 */
#define ROM_BOARD_171_5704		(2)

**Comment by MistyDreams** [comment]
is your port on github?

**Comment by arcadez2003** [comment]
> is your port on github?

Nah the xbox MAME ports were created long before github was even heard of.

**Comment by MistyDreams** [comment]
just the system16 sources thats all i need to look at machine drives and video its just to match the memory maps

**Comment by arcadez2003** [comment]
It's easier to just send the full shebang.......

https://www.sendspace.com/file/y5trtl

**Comment by MistyDreams** [comment]
thanks for that. Once the sound is fixed cottons good to go as well its a lot better than it was though. bullets running fine too. i might just do afighter if its not protected cba doing the analog stuff but the game does work.

**Comment by MistyDreams** [comment]
ps only reaction i got was was 82 in the sound test think it was.

retroarch: symbol lookup error: /home/misty/Source/mame2003-plus-libretro/mame2003_plus_libretro.so: undefined symbol: update_adpcm

**Comment by MistyDreams** [comment]
had a look seems like slave mode is causing the sound loss which your core is set too. if you put memory region in its fine but you loose the samples like you have it. the porting  looks ok  to me only changes that are needed is in the interface ill. Honestly am at a loss with this one the aliensyn works at least plays the right samples but messes up on game over sound has to be something simple we are missing suppose live without samples for now.

**Comment by MistyDreams** [comment]
```
static data16_t unmapped_memory_r(void)
{
	static int recurse = 0;
	data16_t result;

	/* Unmapped memory returns the last word on the data bus, which is almost always the opcode */
	/* of the next instruction due to prefetch; however, since we may be encrypted, we actually */
	/* need to return the encrypted opcode, not the last decrypted data. */

	/* Believe it or not, this is actually important for Cotton, which has the following evil */
	/* code: btst #0,$7038f7, which tests the low bit of an unmapped address, which thus should */
	/* return the prefetched value. */

	/* prevent recursion */
	if (recurse)
		return 0xffff;

	/* read original encrypted memory at that address */
	recurse = 1;
	result = program_read_word_16be(activecpu_get_pc());
	recurse = 0;
	return result;
}
```
thats probable the issue!

**Comment by arcadez2003** [comment]
Is that not for the FD1094 encryption.??

**Comment by arcadez2003** [comment]
Maybe i should tried a couple more things to try and get cotton to play with our original UPD7759 before going down
this path. @MistyDreams here is a copy of your latest WIP driver with the original soundcore hookup with a few extra diffs
would you like to give it a quick whirl..?? it will be the last role of the dice more or less......

I would myself but since you've more or less got this wrapped up i deleted all my WIP cores and fired up LEGO Star Wars
once again :)
[try this.zip](https://github.com/libretro/mame2003-plus-libretro/files/8892072/try.this.zip)

**Comment by MistyDreams** [comment]
ive tried different combos with the core now with the old though as it has a different user interface.  

the code as is doesnt compile like you am done with it will put the old alien3 fix back in that plays the wrong samples as well!

**Comment by arcadez2003** [comment]
Yeah small mistake in that read it should have been

static READ_HANDLER( sys16_upd7759_status_r )
{
    return UPD7759_busy_r(0) << 7;
}

I tried out the changes same as before with the ole soundcore the music and sfx cut out, like you im done now so i guess
we'll use the UPD7759 for every game in system16.c and only use the UPD7759b for Cotton and mark the game as
imperfect sound.

**Comment by MistyDreams** [comment]
well its up to @mahoneyt944 to deal with both games bullet and cotton are working apart from the sample glitches. either silent or garbled. its up to him how the libretro port deals with it

**Comment by mahoneyt944** [comment]
Can't see any harm in using UPD7759b /UPD7759 as separate designations. Ideally this reduces issues and having to test many games. Also the ability to cross test games in the future maybe useful.

**Comment by mahoneyt944** [comment]
PR these games and I'll get these merged into master.

**Comment by MistyDreams** [comment]
cant pr branches are too diverged would need done from scratch

**Comment by MistyDreams** [comment]
very few games need the 7559 testing and converting should be simple enough
```
src/includes/system16.h:359:extern struct UPD7759_interface sys16_upd7759_interface;
src/includes/system16.h:360:extern struct UPD7759_interface aliensyn_upd7759_interface;
src/drivers/twin16.c:945:static struct UPD7759_interface upd7759_interface =
src/drivers/bladestl.c:430:static struct UPD7759_interface upd7759_interface =
src/drivers/mainevt.c:701:static struct UPD7759_interface upd7759_interface =
src/drivers/combatsc.c:656:static struct UPD7759_interface upd7759_interface =
src/drivers/tmnt.c:2437:static struct UPD7759_interface upd7759_interface =
src/drivers/segac2.c:4300:static struct UPD7759_interface upd7759_intf =
src/drivers/88games.c:284:static struct UPD7759_interface upd7759_interface =
src/drivers/prehisle.c:229:static struct UPD7759_interface upd7759_interface =
src/drivers/rpunch.c:608:static struct UPD7759_interface upd7759_interface =
src/drivers/snk68.c:802:static struct UPD7759_interface upd7759_interface =
src/sound/upd7759.c:118:	const struct UPD7759_interface *intf;
src/sound/upd7759.c:284:	const struct UPD7759_interface *intf = msound->sound_interface;
src/sound/upd7759.h:14:struct UPD7759_interface
src/machine/system16_machine.c:2205:struct UPD7759_interface sys16_upd7759_interface =
```

**Comment by mahoneyt944** [comment]
I'm not against converting but I also don't have time for testing right now. Typically that's where using it as "plug and play" for the games that need it is ideal but if either of you want to test every game that used the old sound and verify that the new sound is a good replacement, that's fine too. I have no idea how many of those games may or may not have "hacks" or volume adjustments to suit the old sound. TMNT is a game that I know did have tweaks, because there were user reported issues with volume levels for certain sounds throughout the game. So stuff like that would need sorted and a full playthrough to test for these areas.

As far as branches being too different from master, this is unavoidable...but I'm would recommend doing small incremental changes, PR them, and consistently sync with master before starting the next change. This is also ideal for testing purposes later when things get broken accidentally. We can back track the the commit responsible.

**Comment by MistyDreams** [comment]
hmmm ill add the games at some point the sound hookup you can deal with that up to you how to deal with it.

**Comment by MistyDreams** [comment]
@arcadez2003  put a pr and closed it feel there is too much nit picking pitty cotton is a decent game wasted enough time on this

**Comment by mahoneyt944** [comment]
@MistyDreams
It is a pitty..... You created a PR which moved around initialized variables, which in turn, effects every game in the driver without testing the effect this might have on those games. Your PR created two machine drivers for cotton, one which was imported into the other, which makes no sense. You failed to deliberate about the PR in an open manor and ignored all feedback given. Expecting me or anyone else to finish the work done to a PR you've created is outrageous.

**Comment by MistyDreams** [comment]
> @MistyDreams It is a pitty..... You created a PR which moved around initialized variables, which in turn, effects every game in the driver without testing the effect this might have on those games. Your PR created two machine drivers for cotton, one which was imported into the other, which makes no sense. You failed to deliberate about the PR in an open manor and ignored all feedback given. Expecting me or anyone else to finish the work done to a PR you've created is outrageous.

i did test moving the proc to init has no side effects in know the order of whats happening  proof?

The sound is your choice as maintainer thats why i left the double framework for you so its quicker to implement arcaze2003 done something similar.

i closed the pr easier than expanding explanations to you.

**Comment by mahoneyt944** [comment]
Then why bother even moving it. All the other games are framed a certain way, why not follow that? That way there can be a future commit to just eliminate the machine inits or whatever else.... Make incremental changes that we can check against later if something breaks.

Make one commit to add the games, another to remove machine inits, another to test game by game going to the new drawing system etc etc. Incremental changes.....

**Comment by MistyDreams** [comment]
I never removed any machine inits what are you talking about

**Comment by mahoneyt944** [comment]
Reference every other driver_init in the driver...the code you placed in your driver_inits, is typically in the machine_init for the other games. Which is why you had to move those initialized variables.

My point is to add the games in the format that's currently being used by the driver. Just the games, nothing else.

Then if you want to change the format, do it in another PR. Just the format, nothing else.

If you want to change a game to draw with the new format or something, do it in another PR. Etc etc.

Incremental changes. Which can be tested for faults easily.

**Comment by MistyDreams** [comment]
there are no faults it simple machine init comes before video init all video init does is wipe out the proc setting if you set it machine init no side effects to anything already setup. Your making little sense nothings broke ive tested this on my branch

**Comment by mahoneyt944** [comment]
I don't really care if we use machine_init or driver_init to initialize these, I'm just suggesting that we do all the games the same way so that there is consistency in the driver. I'm not indicating that doing it that way is wrong or breaking things, just that I think these should be in separate PRs. one to add the games and one to switch all the games to use driver init instead of machine init.

If it would be an easier solution, I can update all the games to use driver init then you can re-PR your branch.

**Comment by arcadez2003** [comment]
Right for me less work is always more that's my whole lazy way of looking at things my advice with that in mind is.......

The Sega video code affects multiple drivers and a very high number of games, there has been some lovely work done
backporting newer routines from Segas16B for use here in this core, but surely this video code can only be considered
Wip and experimental at the moment.??

As a result IMO it's best to isolate it just now and only use it for the games that really require it if im not mistaken that
list is quite short eg Action Fighter, Cotton, Fantasy Zone II, Tough Turf and Wrestle War, in time we can test more
of the sys 16 games in depth to see if they benefit from using it and if so shift them across.

With regards to the UPD7759 sound core my advice is to leave the original sound core intact and hooked up to 
every driver that uses it, then either we can fix Cotton so it'll work with it or look at the alternative updated soundcore
i created for my Xbox core just for use in Cotton and no other games unless we're willing to test everyone of them.

@MistyDreams did you say you got Cotton to work with the original sound core.??

**Comment by mahoneyt944** [comment]
I never said anything was broke. I said if things break later we can track it easier.

**Comment by MistyDreams** [comment]
@arcadez here is the pr he is moaning about https://github.com/libretro/mame2003-plus-libretro/pull/1376

**Comment by arcadez2003** [comment]
> yea @arcadez2003 all your need to do is change the interupt line to if (chip) in machine/system16.c in the cause interrupt code but the old code not sure if that breaks any other game though runs out of sync with RA with the old core though

Then i'd isolate it just for Cotton until every game in System 16 has been tested with that change and if we cant be
bothered testing them all just now "i know i cant my health has taken a turn for the worse again" then it's no problem 
really as every system 16 game will be fine in the sound dept in the mean time.

Dont think im nitpicking here or not trusting your changes that's my default position with my own code dont trust it
until fully tested :)

**Comment by MistyDreams** [comment]
oI did isolate cotton he complained aboiut two inits left the poath for him to update the audio see for yourself in the pr above honestly nothing breaking there I left it to him to sort the sound out legt the foundation in so he could edit it as well

**Comment by arcadez2003** [comment]
Right well i've stated my opinons rightly or wrongly depending on your viewpoint i suppose TBH it's something you'll both
have to work out for yourselves TBH as im not really too well today so dont feel like trying to a middle man peacemaker :)

**Comment by MistyDreams** [comment]
its is finished sounds just fucked like every other game that uses the old core

**Comment by MistyDreams** [comment]
made lovely progress with new games sound is workaroundable for cotton thing is fantzn2 and and cotton look great with the split effects fixed like when the rock bounces on the floor and the screen splits who would thought this core was doable for system 16 just goes to show

**Comment by mahoneyt944** [comment]
The wave effect in fantzn2x's bg is pretty cool. Much nicer then the still picture lol

**Comment by MistyDreams** [comment]
yea told you that needed fixed ages ago. ill check if the new code updates the missing attract screen tiles in fantzn2 just for my own curiosity have a little personal version of system16 for my own use

**Comment by mahoneyt944** [comment]
I'm not looking to create hostility here, this core is just a hobby for me. These discussions from earlier went a little to far over the fence, and for that I'll apologize. I hope we can resume a friendly environment going forward.

Anyways,
I ended up updating the type0 textram to include the 2nd layer and went through every game that uses that proc and it seems like it works fine with every game as far as I could tell. Ideally I'd like to start switching games to use the new drawing one by one, I think most should be fine to do so, but I'd like to verify each one during the conversion.

**Comment by MistyDreams** [comment]
Honestly I dont care what you take or not. Im fixing cotton for my nieces they love the pi for some reason. I quiet like this game as well cant believe its one i missed. 

The tile system needs fixed, the the new draw will fix the sprite issues though which there are a few of only for systemb though. 

I spent quiet a bit of time on my core testing this stuff I added and to be honest does no harm to anything else at all. Ive went a different route with the games ive added as the old system isint working well so i updated to a newer one from another mame version its not perfect but will suffice for the games we are using in the core for systemb if any have issues on my end wit my naggy little nieces. If its not broke dont fix it!

**Comment by mahoneyt944** [comment]
I've never played it, don't know anything about it really. Must be decent if you guys are going through hoops to get it right.

**Comment by MistyDreams** [comment]
game is done just samples dont play music and effects are fine

**Comment by mahoneyt944** [comment]
👍

**Comment by mahoneyt944** [comment]
Btw do you see any reference to tturfbl changing it's sprxoffset in later mame? I noticed it's starting point on screen is a few pixels different then the parent. Not by much but it's different. Wondering if the bootleg is just like that or if we need to adjust it

**Comment by MistyDreams** [comment]
not sure mame stopped using that video system we are a long time ago as well most things dont require offsets

**Comment by arcadez2003** [comment]
Oh my look at all this graft
https://github.com/MistyDreams/mame2003-plus-libretro/commit/b1183ef751c79a830a4aab56ae64bd07f4e3e951 
Makes me wonder if your gonna stop at the system16 junction or carry on down the road that leads to segas18 and 
X-Board :)

**Comment by arcadez2003** [comment]
Aye i had a feeling it would be dependent on whether your niece liked Alien Storm and Thunder Blade or not :) 

Anyway your prio changes in tilemap.c have caught my eye as they could be used the sort some gfx niggles in the
Kaneko16 and Namco System 1 games, i nevered dared to go down this road myself as well my glass half full nature
worried that i might mess it up seeing as every game will be affected.

But i noticed you just tend to charge right in there which is likely the mindset to have for this coding malarky......


0.102u1: Several fixes for the Kaneko16 driver [Pierpaolo Prazzoli]: Fixed tilemap priorities. Fixed some sprites priorities. Added some missing coin counters. Added right palette filling for games using sprite type 1. Swapped berlwall and berlwalt sets as per SebV notes in the source code.

0.89: Namco System 1 update [Nicola Salmoria]. Added a new function tilemap_draw_primask(). This takes an additional parameter, which is ANDed with the contents of priority_bitmap before ORing it with the specified priority. Essentially, the old tilemap_draw() is equivalent to tilemap_draw_pribuffer(..., 0xff). The two should probably be unified. This new handling of priority_bitmap allows a more intuitive handling of pdrawgfx(), and more layers to be used (up to 31, instead of just 5 - Namco S1 needed 8). NOTE: To implement this change, I had to disable the asm optimized routines PDT16 and PDO16 in windows\oslinline.h. If they are useful for performance, they should be updated and reenabled. Fixed sprite handling and removed all the nonsense hacks from the video update. Made sprite/tilemap priorities orthogonal (not sure if it was needed). Vastly improved the key custom emulation, shrinking it down to three basic types. This also fixed issues with pistoldm (e.g. bomb explosions, and other random events). Cleaned up input ports (using PORT_INCLUDE). Found some more debug dipswitches. Added support for visibility window, this fixes countless issues with partial video updates and the intro of Rompers. Better ROM loading, closer to the hardware (though it uses more memory). Toggling the test mode dipswitch is enough to enter test mode - no need to reset. Cleaned up the CUS30 wavedata/sound command/RAM interface, put it all in a single function. Changed M6809 CPU1/2 clock speeds to 1536000 Hz, palettesize to 8192 colors, VSync to 60.606060 Hz and Namco sound to Namco CUS30. Fixed user1 rom loading.

**Comment by arcadez2003** [comment]
@MistyDreams i meant to ask ya what version of MAME did you take the Segas16a and 16b video code from.?? reason i
ask is MAME90 versions of these are quite a bit better performance wise than MAME91 onwards which can and will make
a difference on the devices which use this core.

**Comment by MistyDreams** [comment]
The code was drop in added 3 lines to it literally to hook the textram tileram to the segaic suprised no one done it before! Anyway i tested batman and some deco games they make heavy use of the tilemaps. It probably wont see the light of day bendy two moany little kids sitting in front of there pie.

**Comment by arcadez2003** [comment]
Aye your work has got me thinking off when i undertook which for me was the mammoth task of updating the Sega code
while keeping the older version of it intact well i was fitter 7 years back :)

Updated M68K CPU updated memory handling 3 seperate Sega cores three seperate video routines two seperate machine routines double the amount of drivers updated tilemap and an extra sound core to top it all off if only you were there to advise me all those years ago as it looks like it could have been done with a quarter of the hastle :)

**Comment by MistyDreams** [comment]
hindsight's a great thing this doesnt effect the other system16 driver well apart from commenting out fanzn2x in the game definition at the bottom leaves options to add games at some point.

**Comment by arcadez2003** [comment]
> The code was drop in added 3 lines to it literally to hook the textram tileram to the segaic suprised no one done it before! Anyway i tested batman and some deco games they make heavy use of the tilemaps. It probably wont see the light of day bendy two moany little kids sitting in front of there pie. The codes free for anyone to take don't wanna deal with pull requests though find them tedious

Not trying to be funny here but are the tilemap changes really required.?? i dont know the in's and out's off all the changes
you made but i dont remember having to touch the tilemaps whatsoever when i did the aforementioned update other
than this call for the Y-Board games i think.....

void *user_data;		/* user-supplied tilemap-wide pointer */

**Comment by MistyDreams** [comment]
for me personally yes dont mind what other people do though. Just on my personal branch so done mind how anyone rips it up

**Comment by arcadez2003** [comment]
> for me personally yes dont mind what other people do though. Just on my personal branch so done mind how anyone rips it up

Yeah fair enough personally i like to add the code test everything connected to it and when im happy it's all good then im
ready to share it out to the mainline, i dont wanna bang on about my health it isn't what it was that's just the way it goes
so as a result i cannot really take on the big jobs anymore so i have no plans to look at your branch changes with a view
to porting them across to here.

My advice take your time work through everything and when your happy surely it's time to make a pull req and send it
to mainline and that way loads of other people can take advantage of your graft here and play some really nice games
they previously could not and well it'll be you they can thank for that.

But at the end of the day it's upto yourself here what you choose to do as per past events im only an outside contributor
nowadays so im in a similar situation as you in the sense i can come and make my case here with regards to any commits i 
might wanna make but at the end of the day the final decisions lie with the maintainers good ole fashioned give and take
will usually get the job done ;)

**Comment by MistyDreams** [comment]
Yea well its up to the maintainers what they want and dont. What I dont want to get into a situation where you debating more than coding when it gets to that stage the maintainer can just grab what they want instead of expecting you to dance to there expectations that way they add the code the way they want too.

**Comment by arcadez2003** [comment]
Yeah all you can do is make your case as per any changes you wanna make and then the maintainers can weigh up the 
pros and cons then decide to use em or not it's simple as that really, with small changes it's usually hastle free as i recall 
it's when i wanted to make the big ones is when the trouble started hence we now have two MAME2003 cores.

You can take this however you want im not gonna really argue with ya as i have no care to do so i appreciate the job 
you've done on these games but all i'll say is a huge change to the tilemap handling which will effect every game 
potentially in this core needs more care and attention than to say i tested some a handful of games in a couple of 
drivers.

The game breakage potential here is very high that's not to say there will be any but rememeber MAME is and was 
fluid in the sense they break and fix stuff with every release there might be issues as per this change which will need 
addressed down the road there might not be time will tell, but really should you not take ownership with regards to 
pushing these commits and fixing any issues which may or may not arise yourself.?? 

As i said im not a maintainer anymore with regards to this core so im not gonna tell ya what to do here or debate
what you should or shouldn't do that is for those in the positions of power to iron out between ya's personally
i'd like to see these extra games playable and im sure it'll be a real boon for the many users of this core to have
em up and running at the end of the day you either port the code across yourself or you dont.

Right i've had my say it might not be popular but i feel it had to be said my work here is done for the moment
im back playing Lego Star Wars on my switch and then it's gonna be Xenoblade Chronicles 3 i might return
sometime to contribute a little something i may not but either way i wish ya's well with regards to this core
later folks it's been a blast.

**Comment by MistyDreams** [comment]
I dont have any view on it at all though can use what they like tilemap changes or not(i plan on making use of the changes personally). 

Im not taking ownership of something I done for myself will share it sure no bother there. Adding the tilemap changes or not are neither here to me maintainers can do as they please. This code effects nothing else if you leave the tilemap changes. Let the maintainers do there thing. Im not getting into doing a maintainers job like drama free events if possible. I wouldn't fancy a maintainers job not worth the hassle.

**Comment by mahoneyt944** [comment]
We just have to be delicate with each PR so we can trace back issues that may arise later. That's all.

For example with your branch:
https://github.com/MistyDreams/mame2003-plus-libretro/tree/test_new_soundcore

Submit this as a single PR, once it's ready of course. Then we can merge it and start testing lots of games. If we find any breakage we'll create followup PRs to fix it until the core is sound again, no pun intended.

This branch is your work, not in the sense of ownership, but the time taken to switch out things in this core and the test results of doing so. If it's going to be added here, you'll have to make a PR for it. If you don't want to make a PR, that's ok too but I don't plan to scrape through test branches and determine if they are ready or not.

**Comment by MistyDreams** [comment]
yea we understand each other @mahoneyt944  again im working on my own terms and agenda. Thars testing the sound in other games see if i can get cotton fixed. My goal isint to do pr for here .

**Comment by mahoneyt944** [comment]
That's totally fine, however, the collaboration here is for this repository. If you are only interested in your own agenda and progress of your own fork then you can have those discussions in your personal repository as it clouds the issue for those who wish to work on the codebase here. This is the entire reason for having forked projects in the first place.

**Comment by arcadez2003** [comment]
> Its not a personal fork its fixing a issue for my nieces but if you want to get political I get it. Im not interested in being forced into things that adds a workload to my life not of my choosing.

Right I want to see Cotton playable in this core so i'll make you an offer here when you finish up making all the changes and 
are happy that the work is complete and if you want send me the code sometime and i'll port it across to MAME2003+.

But i cant guarantee that i wont just cherrypick mix and match more or less to get the game up and running here with the 
most minimal changes required to the mainline codebase in order to do so.

Regards

**Comment by MistyDreams** [comment]
never made any serious changes to main code the ironic thing is the the initialization changes mahony didnt want he cherry picked and left cotton out of the commit. Even added this missing spacing one. Ill sort something for you and leave the zip later on today will be easy enough to add it just be warned there is gfx glitches with current tiling code

**Comment by mahoneyt944** [comment]
I'm actually overhauling the init structuring now in a new direction.

**Comment by MistyDreams** [comment]
ok let me know when your done so i can sort a fix for arcadez

**Comment by arcadez2003** [comment]
> I dont have any view on it at all though can use what they like tilemap changes or not(i plan on making use of the changes personally).
> 
> Im not taking ownership of something I done for myself will share it sure no bother there. Adding the tilemap changes or not are neither here to me maintainers can do as they please. This code effects nothing else if you leave the tilemap changes. Let the maintainers do there thing. Im not getting into doing a maintainers job like drama free events if possible. I wouldn't fancy a maintainers job not worth the hassle.

It's not drama to point out that a significant change to the tilemap code affects every game that uses that drawing method
it's a simple difference of opinion im stating you dont need to make this change for the Sega games i didn't and they're
all fine in my Xbox core.

And you think it's required that's fine everyone is entitled to their own viewpoint all im saying is we have to be careful here......

 [Nicola Salmoria]. Added a new function tilemap_draw_primask(). This takes an additional parameter, which is ANDed with the contents of priority_bitmap before ORing it with the specified priority. Essentially, the old tilemap_draw() is equivalent to tilemap_draw_pribuffer(..., 0xff). The two should probably be unified. This new handling of priority_bitmap allows a more intuitive handling of pdrawgfx(), and more layers to be used (up to 31, instead of just 5 - Namco S1 needed 8). NOTE: To implement this change, I had to disable the asm optimized routines PDT16 and PDO16 in windows\oslinline.h. If they are useful for performance, they should be updated and reenabled.

You've deleted your fork again so i cant check to see if you've implemented all of the changes above or just in part.??
anyway at the end of the day regardless of this core you need to be carefull with the build you create for your relatives
as you dont wanna break a whole load of games just to get one playable :)

**Comment by MistyDreams** [comment]
@arcadez2003 my forks not deleted the tilemap functions are for my own personal use in the future. I at no pint said they where needed and I never put a pull request in. Ill add the game to the current framework when mahoney is done.

**Comment by arcadez2003** [comment]
@MistyDreams thanks for the code however i'll need to see what path @mahoneyt944 wants to go down here with regards
to the tilemaps.?? i've already stated what route i'd take so no point going over ole ground.

Anyway i wont have much time to look at this just now as im bust trying to get 100% on each chapter in LEGO Star Wars :)
so maybe by the time i plan to look at this it'll already be done.

**Comment by MistyDreams** [comment]
@arcadez2003 not sure what issue is the codes the same. Just leave the tilemap.c and tilemap.h changes. Does not effect the rest of the code for the change. Seems thats what you have issues with.

**Comment by arcadez2003** [comment]
Ach maybe im overly cautious i dont like big changes hence my whole dev is based around avoiding em but hey with an
attitude like mine we'd never have discovered the Americas as i'd likely have said let's not go incase we dont fucking find it :)

**Comment by MistyDreams** [comment]
its a none issue honestly just leave the tilemap.c and tilemap.h changes out. I did say this was for my personal use and left the code zipped pretend them files arent there and there is no issue apart from adding system16 tile rendering in a different driver

**Comment by mahoneyt944** [comment]
@arcadez2003 @MistyDreams 

Ok so here's my proposal. It seems redundant to call `machine_init_sys16_onetime()` in every driver init. Ideally default variables should be set then be modified by the driver init and /or machine init. The order of calls are as follows when firing up a game:
- driver init
- video start
- machine init

Which means machine init washes any variable set in driver init and video start which they have in common and which is not under restriction.

My thoughts are to move this call earlier where it would precede driver_init and video_start (basically a "video init"). The current PR #1380 does just that, but does not isolate any variables conflicted between driver init, video start, and machine init. That would need to be done in another PR. This should not change any game currently hooked up since `machine_init_sys16_onetime()` was always called first in ever driver init and now it will still be called before driver init

I'd like to, in a followup PR, evaluate the variables set within video start, it seems to reason that these could be moved into sys16_onetime which would allow the driver init to modify any variables without any issue of confliction before video start occurs.

**Comment by MistyDreams** [comment]
driver init
video start init
machine init

order is true but all come before video update anyway.

machine init is the only one that will be called on reset thats why I used that one. In terms of this the order doesnt really matter as long as its set one time.

**Comment by mahoneyt944** [comment]
I agree nothing should be set in video start, I think with after this PR we just move that stuff into sys16_onetime then we'll have the flexibility to set game specific variables before or after video start using driver init vs machine init. Whatever we deem most applicable. Not that it should matter as you noted

**Comment by MistyDreams** [comment]
@arcadez2003 ive been testing a bug in the soundcore for tmnt the thing is the data ive traced in the upd7759 looks correct and the sample data is  there. 

its just not coming out for some reason think its being lost somewhere in the sound system. Can you start tmnt and when it says fire on the screen the sample should play  and say fire (it doesnt on this core the rest are ok though) as i said the data look valid that the upd7759 is spitting out so i think the issues for this is elsewhere. If you confirm that sample plays as wine sound doesn't work early mame version and I dont have windows installed to test it there.

**Comment by mahoneyt944** [comment]
fire plays with the current master.

**Comment by MistyDreams** [comment]
@mahoneyt944 new sound core not the old one

**Comment by mahoneyt944** [comment]
Ohh my bad. Trying to be helpful. I have that game on my personal rom list. Use to love playing the home port on Genesis.

**Comment by arcadez2003** [comment]
> @arcadez2003 ive been testing a bug in the soundcore for tmnt the thing is the data ive traced in the upd7759 looks correct and the sample data is there.
> 
> its just not coming out for some reason think its being lost somewhere in the sound system. Can you start tmnt and when it says fire on the screen the sample should play and say fire (it doesnt on this core the rest are ok though) as i said the data look valid that the upd7759 is spitting out so i think the issues for this is elsewhere. If you confirm that sample plays as wine sound doesn't work early mame version and I dont have windows installed to test it there.

Well that's the thing i only use the newer version of the UPD7759 with Segas16b.c in my xbox port all other drivers tmnt
included use the ole one, but if you want me to test this i can always mod tmnt to use the newer soundcore tomorrow
sometime.

**Comment by MistyDreams** [comment]
its ok for now found the issue for that all other games testes well and that issue is sorted Just need to try get aliensyn working in this test if that works then cotton should it doesnt at the moment

**Comment by MistyDreams** [comment]
good news alien syndrom is fixed with the new core was bad roms

**Comment by mahoneyt944** [comment]
👍 nice

**Comment by MistyDreams** [comment]
cotton samples are still silent think its that unmapped memory read address

**Comment by MistyDreams** [comment]
@arcadez2003  do me a favor in line 458 of your port in sega16c.c

```
static data16_t unmapped_memory_r(void)
{
	static int recurse = 0;
	data16_t result;

	/* Unmapped memory returns the last word on the data bus, which is almost always the opcode */
	/* of the next instruction due to prefetch; however, since we may be encrypted, we actually */
	/* need to return the encrypted opcode, not the last decrypted data. */

	/* Believe it or not, this is actually important for Cotton, which has the following evil */
	/* code: btst #0,$7038f7, which tests the low bit of an unmapped address, which thus should */
	/* return the prefetched value. */

	/* prevent recursion */
	if (recurse)
		return 0xffff;

	/* read original encrypted memory at that address */
	recurse = 1;
	result = program_read_word_16be(activecpu_get_pc());
	recurse = 0;
	return result;
}
```

this only runs one time anyway change 
```result = program_read_word_16be(activecpu_get_pc());```  too ``` return 0xffff; ``` and see if you loose sound

**Comment by arcadez2003** [comment]
With that change the music sfx and speech work fine the game however is borked you die right away on level one start
and then when you respawn your stuck more or less the game doesn't move eg scroll but the enemies still scroll in.

**Comment by MistyDreams** [comment]
ok cheers for that is appreciated just cant get mame89 to run on line wine with sound to checked save me a bit of time. See i had that issue when i returned a value when i returned nothing it was fine ill look into it a little at some point thin we need 24bew for the 68k if im reading the source right

**Comment by MistyDreams** [comment]
see if the ports spitting the data out i suppose was a pain debugging this with all the shite logged with core for tmnt

**Comment by arcadez2003** [comment]
program_read_word_16be = cpu_readmem24bew_word for our core

**Comment by MistyDreams** [comment]
I dont see any customs for cotton in that code we are decrypted already but that check it still there we defo getting data sent through the fifo port and the bank switches are happening

**Comment by MistyDreams** [comment]
add this https://github.com/MistyDreams/mame2003-plus-libretro/commit/802637613fa551a6bebb1ed5d84b3dbd43fa86cc and get some screething now

**Comment by MistyDreams** [comment]
that was a ballache sorted it

**Comment by MistyDreams** [comment]
think we can close this one is solved

**Comment by arcadez2003** [comment]
What an odyssey eh but well done you never gave up and got there in the end.

**Comment by MistyDreams** [comment]
usually do if you trace everything good news is the soundcore is good for all other games and is tested.

**Comment by mahoneyt944** [comment]
Closed, but we can start another if need be.
